base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -1248,69 +1264,45 @@ fn find_strongest_orphan_tip<T: BlockchainBackend>(
-) -> Result<(), ChainStorageError>
-{
-    if db.get_orphan_count()? > orphan_storage_capacity {
-        trace!(
-            target: LOG_TARGET,
-            "Orphan block storage limit reached, performing simple cleanup.",
-        );
-        let mut min_height: u64 = u64::max_value();
-        let mut remove_hash: Option<BlockHash> = None;
-        db.for_each_orphan(|pair| {
-            let (_, block) = pair.unwrap();
-            if block.header.height < min_height {
-                min_height = block.header.height;
-                remove_hash = Some(block.hash());
-            }
-        })
-        .expect("Unexpected result for database query");
-        if let Some(hash) = remove_hash {
-            trace!(target: LOG_TARGET, "Discarding orphan block ({}).", hash.to_hex());
-            remove_orphan(db, hash)?;
-        }
-    }
-    Ok(())
-}
-
 // Perform a comprehensive search to remove all the minimum height orphans to maintain the configured orphan pool
-// storage limit.
-fn cleanup_orphans_comprehensive<T: BlockchainBackend>(
+// storage limit. If the node is configured to run in pruned mode then orphan blocks with heights lower than the horizon
+// block height will also be discarded.
+fn cleanup_orphans<T: BlockchainBackend>(
     db: &mut RwLockWriteGuard<T>,
     orphan_storage_capacity: usize,
 ) -> Result<(), ChainStorageError>
 {
     let orphan_count = db.get_orphan_count()?;
-    if orphan_count > orphan_storage_capacity {
+    let num_over_limit = orphan_count.saturating_sub(orphan_storage_capacity);
+    if num_over_limit > 0 {
         trace!(
             target: LOG_TARGET,
-            "Orphan block storage limit reached, performing comprehensive cleanup.",
+            "Orphan block storage limit reached, performing cleanup.",
         );
-        let remove_count = orphan_count - orphan_storage_capacity;
 
         let mut orphans = Vec::<(u64, BlockHash)>::with_capacity(orphan_count);
         db.for_each_orphan(|pair| {
-            let (_, block) = pair.unwrap();
-            orphans.push((block.header.height, block.hash()));
+            let (block_hash, block) = pair.unwrap();
+            orphans.push((block.header.height, block_hash));
         })
         .expect("Unexpected result for database query");
         orphans.sort_by(|a, b| a.0.cmp(&b.0));
 
+        let metadata = db.fetch_metadata()?;
+        let horizon_height = metadata.horizon_block(metadata.height_of_longest_chain.unwrap_or(0));
         let mut txn = DbTransaction::new();
-        for i in 0..remove_count {
-            trace!(
-                target: LOG_TARGET,
-                "Discarding orphan block ({}).",
-                orphans[i].1.to_hex()
-            );
-            txn.delete(DbKey::OrphanBlock(orphans[i].1.clone()));
+        let mut removed_count: usize = 0;
+        for (height, block_hash) in orphans {
+            if (height > horizon_height) & (removed_count >= num_over_limit) {
```

Comment:
```suggestion
            if height > horizon_height && removed_count >= num_over_limit {
```

Hah... using bitwise `&` on two boolean operands (therefore brackets are needed) and using a boolean `&&` on boolean conditions is the same thing, interesting (though boolean `&&` is more idiomatic)

Reply:
Change applied.